pkg/controller/datavolume-controller.go

```diff
@@ -2599,3 +2623,288 @@ func GetRequiredSpace(filesystemOverhead float64, requestedSpace int64) int64 {
+func getSizeFromAnnotations(sourcePvc *corev1.PersistentVolumeClaim) (string, bool) {
+	virtualImageSize, available := sourcePvc.Annotations[AnnVirtualImageSize]
+	if available == true {
+		sourceCapacity, available := sourcePvc.Annotations[AnnSourceCapacity]
+		currCapacity := sourcePvc.Status.Capacity
+		// Checks if the original PVC's capacity has changed
+		if available && currCapacity.Storage().Cmp(resource.MustParse(sourceCapacity)) == 0 {
+			return virtualImageSize, true
+		}
+	}
+
+	return "0", false
+}
+
+// getSizeFromPod attempts to get the image size from a pod that directly obtains said value from the source PVC
+func (r *DatavolumeReconciler) getSizeFromPod(sourcePvc *corev1.PersistentVolumeClaim) (string, error) {
+	// The pod should not be created until the source PVC has finished the import process
+	if !hasPvcSucceeded(sourcePvc) {
+		return ImgSizeNotReady, nil
+	}
+
+	pod, err := r.createSizeDetectionPod(sourcePvc)
+	if err != nil {
+		return ImgSizeNotReady, err
+	} else if !hasPodSucceeded(pod) {
+		return ImgSizeNotReady, nil
+	}
+	// Parse raw image size from the pod's termination message
+	imgSize := pod.Status.ContainerStatuses[0].State.Terminated.Message
+	err = r.client.Delete(context.TODO(), pod)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return imgSize, err
+	}
+	// Update Source PVC annotations
+	if err := r.updateClonePVCAnnotations(sourcePvc, imgSize); err != nil {
+		return imgSize, err
+	}
+
+	return imgSize, nil
+}
+
+// createSizeDetectionPod handles the creation of the size-detection pod
+func (r *DatavolumeReconciler) createSizeDetectionPod(sourcePvc *corev1.PersistentVolumeClaim) (*corev1.Pod, error) {
+	resourceRequirements, err := GetDefaultPodResourceRequirements(r.client)
+	if err != nil {
+		return nil, err
+	}
+
+	podName := sizeDetectionPodName(sourcePvc)
+	pod := &corev1.Pod{}
+	nn := types.NamespacedName{Namespace: sourcePvc.Namespace, Name: podName}
+	// Trying to get the pod if it already exists/create it if not
+	if err := r.client.Get(context.TODO(), nn, pod); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return nil, err
+		}
+		// Generate the pod spec
+		pod = makeSizeDetectionPodSpec(sourcePvc, r.client, r.pullPolicy)
+		if pod == nil {
+			return nil, errors.Errorf("Size detection pod could not be generated.")
+		}
+		if resourceRequirements != nil {
+			pod.Spec.Containers[0].Resources = *resourceRequirements
+		}
+		// Create the pod
+		if err := r.client.Create(context.TODO(), pod); err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return nil, err
+			}
+		}
+
+		r.log.V(3).Info("Size-detection pod created\n", "pod.Name", pod.Name, "pod.Namespace", pod.Namespace)
+	}
+
+	return pod, nil
+}
+
+// makeSizeDetectionPodSpec creates and returns the full size-detection pod spec
+func makeSizeDetectionPodSpec(sourcePvc *corev1.PersistentVolumeClaim, client client.Client, pullPolicy string) *corev1.Pod {
```

Comment:
Any reason you didn't make this part of the reconciler? that way you don't have to pass in the client, and if you follow my below suggestion, you don't have to pass in the image name either.

Reply:
I'll change it :+1: Sometimes I've struggled to guess when I should make a function part of the reconciler, or when I should just pass the client as an argument. Are there any guidelines for this?